apps/lib/opt.c

```diff
@@ -367,14 +367,29 @@ void print_format_error(int format, unsigned long flags)
     (void)opt_format_error(format2str(format), flags);
 }
 
+/*
+ * Bump count of provider options, return new value.
+ * This function should be cleaner, and be in app_provider.c, but
+ * the tests only link against this file so it has to be here and
+ * this API is clunky but workable.
+ */
+int opt_incr_provider_options(int x)
+{
+    static int count;
+
+    return count += x;
+}
+
 /* Parse a cipher name, put it in *EVP_CIPHER; return 0 on failure, else 1. */
 int opt_cipher_silent(const char *name, EVP_CIPHER **cipherp)
 {
     EVP_CIPHER_free(*cipherp);
 
     ERR_set_mark();
     if ((*cipherp = EVP_CIPHER_fetch(NULL, name, NULL)) != NULL
-        || (*cipherp = (EVP_CIPHER *)EVP_get_cipherbyname(name)) != NULL) {
+            || (!opt_incr_provider_options(0)
+                && (*cipherp =
+                        (EVP_CIPHER *)EVP_get_cipherbyname(name)) != NULL)) {
         ERR_pop_to_mark();
         return 1;
     }
```

Comment:
Not sure why the counter is needed...  I would have expected something like this:

``` C
int opt_cipher_silent(const char *name, EVP_CIPHER **cipherp)
{
    OSSL_LIB_CTX *libctx = app_get0_libctx();
    const char *propq = app_get0_propq();

    EVP_CIPHER_free(*cipherp);

    
    ERR_set_mark();
    if ((*cipherp = EVP_CIPHER_fetch(libctx, name, propq)) != NULL
        || ((libctx == NULL && propq == NULL)
            && (*cipherp = (EVP_CIPHER *)EVP_get_cipherbyname(name)) != NULL)) {
        ERR_pop_to_mark();
        return 1;
    }
    ERR_clear_last_mark();
    return 0;
}
```

Reply:
> Not sure why the counter is needed... I would have expected something like this:

It's really a flag, but I just did ++ instead of `= 1`  And it's a function because as you pointed out elsewhere, sharing global variables is messy.  More importantly, if there are config files loaded that set default, I wasn't sure about what would happen.  (The whole interaction of libctx and default providers is murky to me.)

Finally, I didn't want to set the libctx/propq as local variables because an errors woudl be outside the error stack clearing from the `set-mark` function.